src/goto-instrument/code_contracts.cpp

```diff
@@ -282,9 +354,345 @@ const symbolt &code_contractst::new_tmp_symbol(
+          called_op_it != called_assigns.operands().end();
+          called_op_it++)
+      {
+        if(
+          freely_assignable_exprs.find(*called_op_it) !=
+          freely_assignable_exprs.end())
+        {
+          continue;
+        }
+        exprt alias_expr =
+          create_alias_expression(assigns, *called_op_it, aliasable_references);
+
+        goto_programt alias_assertion;
+        alias_assertion.add(
+          goto_programt::make_assertion(alias_expr, ins_it->source_location));
+        program.insert_before_swap(ins_it, alias_assertion);
+        ++ins_it;
+      }
+    }
+  }
+}
+
+bool code_contractst::check_for_looped_mallocs(const goto_programt &program)
+{
+  // Collect all GOTOs and mallocs
+  std::vector<goto_programt::instructiont> back_gotos;
+  std::vector<goto_programt::instructiont> malloc_calls;
+
+  int idx = 0;
+  for(goto_programt::instructiont ins : program.instructions)
+  {
+    if(ins.is_backwards_goto())
+    {
+      back_gotos.push_back(ins);
+    }
+    if(ins.is_function_call())
+    {
+      code_function_callt call = ins.get_function_call();
+      const irep_idt &called_name =
+        to_symbol_expr(call.function()).get_identifier();
+
+      if(std::strcmp(called_name.c_str(), "malloc") == 0)
+      {
+        malloc_calls.push_back(ins);
+      }
+    }
+    idx++;
+  }
+  // Make sure there are no gotos that go back such that a malloc is between
+  // the goto and its destination (possible loop).
+  for(auto goto_entry : back_gotos)
+  {
+    for(const auto &t : goto_entry.targets)
+    {
+      for(auto malloc_entry : malloc_calls)
+      {
+        if(
+          malloc_entry.location_number >= t->location_number &&
+          malloc_entry.location_number < goto_entry.location_number)
+        {
+          log.error() << "Call to malloc at location "
+                      << malloc_entry.location_number << " falls between goto "
+                      << "source location " << goto_entry.location_number
+                      << " and it's destination at location "
+                      << t->location_number << ". "
+                      << "Unable to complete instrumentation, as this malloc "
+                      << "may be in a loop." << messaget::eom;
+          throw 0;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+bool code_contractst::add_pointer_checks(const std::string &func_name)
+{
+  // Get the function object before instrumentation.
+  auto old_fun = goto_functions.function_map.find(func_name);
+  if(old_fun == goto_functions.function_map.end())
+  {
+    log.error() << "Could not find function '" << func_name
+                << "' in goto-program; not enforcing contracts."
+                << messaget::eom;
+    return true;
+  }
+  goto_programt &program = old_fun->second.body;
+  if(program.instructions.empty()) // empty function body
+  {
+    return false;
+  }
+
+  const irep_idt func_id(func_name);
+  const symbolt &f_sym = ns.lookup(func_id);
```

Comment:
`function_symbol`. This is one the problems with abbreviations aside from the obvious (that they hurt readability), I count already 3 different abbreviations for the word ‘function’ in this function alone. Avoid them.

Reply:
Done.